### ISSUE TITLE
"The node <x> already exists in the voters set" error when creating a RawNode

### ISSUE THREAD

**Issue by henrytill** [pingcap/raft-rs#170]
**Describe the bug**
We often see errors like the following when we create instances of `RawNode`:
```
ERROR 2019-01-23T18:43:20Z: raft::raft: The node 1 already exists in the voters set.
```

**To Reproduce**
1. Create a file `examples/two_mem_nodes/main.rs`:
```rust
extern crate env_logger;
extern crate log;
extern crate raft;

use raft::prelude::*;
use raft::storage::MemStorage;

fn make_raw_node(configs: &[Config], idx: usize) -> Result<RawNode<MemStorage>, raft::Error> {
    let peers: Vec<Peer> = configs[idx]
        .peers
        .iter()
        .map(|id| Peer {
            id: *id,
            ..Default::default()
        })
        .collect();
    RawNode::new(&configs[idx], MemStorage::new(), peers)
}

fn main() {
    let _ = env_logger::try_init().unwrap();

    let configs = vec![
        Config {
            id: 1,
            peers: vec![2, 3],
            ..Default::default()
        },
        Config {
            id: 2,
            peers: vec![1, 3],
            ..Default::default()
        },
        Config {
            id: 3,
            peers: vec![1, 2],
            ..Default::default()
        },
    ];

    let _raw_node_1 = make_raw_node(&configs, 0).unwrap();
    let _raw_node_3 = make_raw_node(&configs, 1).unwrap();
    let _raw_node_3 = make_raw_node(&configs, 2).unwrap();
}
```

2. Update `Cargo.toml` accordingly:
```diff
diff --git a/Cargo.toml b/Cargo.toml
index 7ed554f..f2146af 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -38,3 +38,7 @@ travis-ci = { repository = "pingcap/raft-rs" }
 [[example]]
 name = "single_mem_node"
 path = "examples/single_mem_node/main.rs"
+
+[[example]]
+name = "two_mem_nodes"
+path = "examples/two_mem_nodes/main.rs"

```

3. Run it:
```
$ cargo run --example two_mem_nodes -- --nocapture
   Compiling raft v0.4.0 (/home/ht/src/pyrofex/other/raft-rs)                                                                      
    Finished dev [unoptimized + debuginfo] target(s) in 2.33s                                                                      
     Running `target/debug/examples/two_mem_nodes --nocapture`
ERROR 2019-01-23T20:07:33Z: raft::raft: The node 2 already exists in the voters set.
ERROR 2019-01-23T20:07:33Z: raft::raft: The node 3 already exists in the voters set.
ERROR 2019-01-23T20:07:33Z: raft::raft: The node 1 already exists in the voters set.
ERROR 2019-01-23T20:07:33Z: raft::raft: The node 3 already exists in the voters set.
ERROR 2019-01-23T20:07:33Z: raft::raft: The node 1 already exists in the voters set.
ERROR 2019-01-23T20:07:33Z: raft::raft: The node 2 already exists in the voters set.
```

**Expected behavior**
I expect to see no errors when creating an instance of `RawNode`.

**System information**
N/A

**Additional context**
It *appears* that doing the following is redundant, as the node's `ProgressSet` is already initialized properly when we call `RawNode::new` (by virtue of calling `Raft::new`):
https://github.com/pingcap/raft-rs/blob/c6b5d57dbc172dd50f94b0d9e4df2684338dab92/src/raw_node.rs#L253-L255

However, simply removing this causes some of the tests to fail, so I defer to your expertise on how to best address this problem.

**Comment by Fullstop000** [comment]
The field `peers` in `Config` should be used only for convenience when testing because those `peers`  will become `Progress` directly without any `ConfChange` propose . See the comment
```rust
    /// The IDs of all nodes (including self) in
    /// the raft cluster. It should only be set when starting a new
    /// raft cluster.
    /// Restarting raft from previous configuration will panic if
    /// peers is set.
    /// peer is private and only used for testing right now.
    pub peers: Vec<u64>,
```
I suggest just using peers in `Vec<Peer>` passed to `RawNode::new` for a complete `AddNode` mechanism.

Hope the explanations above can be helpful :) @henrytill

**Comment by henrytill** [comment]
@Fullstop000: thanks for your suggestion.

I must admit that I find the content of that comment a bit ambiguous in light of your comment.  It does say:
```
/// It should only be set when starting a new raft cluster.
```
which is what I'm trying to do here.  

Furthermore, the `for` loop I highlighted above is supposed to run when `last_index == 0`, which is also what I'd expect it to be when starting a new cluster.  OTOH, if I was starting a cluster with a non-empty store, where `last_index` was greater than 0, it would skip that loop that causes the error altogether.

**Comment by Fullstop000** [comment]
Thanks for pointing out that the ambiguous comment. 
I truly agree that it should be more clear . For now `peers` and `learners` in `Config` are `pub` fields by some testing reasons but they should be private . You can see the same implement in [etcd-raft](https://github.com/etcd-io/etcd/blob/master/raft/raft.go#L121) that both fields are private . 

About the `last_index` you mentioned, let's dig a little deeper : 

When you start a cluster from scratch (`last_index` is 0) by creating `RawNode` with both `Config.peers` set and passing `peers`explicitly to `RawNode::new` , the error occurs because the  peers will overlap with the `ProgressSet` after `Config.peers` are directly added into it.

If you starting a cluster with non-empty store ( `last_index` is not 0 )  and `Config.peers` is set , panic happens in this situation because raft tries to recover all the progresses from `Storage` but we still has `Config.peers`. The reason of this implement is that `ConfChange`s can happen ( remove or add node ) in a running raft cluster but `Config.peers` is something static. 
https://github.com/pingcap/raft-rs/blob/24280e8f0f9e9a38d8aa62cfb42d1f2f3cf0a026/src/raft.rs#L216-L228
Raft will store the real time cluster configuration in snapshot (`ConfState`) so the node should read it from snap every rebooting which explains why implement skips the peer adding loop you post above. 

Sorry for the verbose explaining. 
TLDR: If you want to a complete raft cluster initializing, do not set `peers` or `learners` in `Config` .

**Comment by Hoverbear** [comment]
An update to this, you can see that in https://github.com/pingcap/raft-rs/pull/196/files#diff-3e643971842b1203495f3e09ab4b6832L222 we removed the peers param from `RawNode::new()`.

Prior to the full 0.6.0 release we'll need to make sure we have a clear correct way documented for folks to bootstrap and restore a cluster.
